### PR TITLE
Add method for computableSubgrids

### LINKED ISSUE #124: Extend documentation on ContainerUtils.buildSubGridContainers(...) + add method for slacks in SubGridContainer
This refers to #66 

Currently it is not clear documented, what happens with the transformer nodes (A,B for two winding transformers and A,B,C,INTERNAL for three winding transformers) in respect to the Resulting `SubGridContainer`. 

After a first, fast investigation _it seems_ that all real nodes (A,B,C) are _copied_ in each subnet and will be added to the nodes set in the resulting `RawGrid`.

This makes sense from my point of view but should be documented somewhere (maybe in the method javadoc?). 

I would suggest the following documentation:

```
 /**
   * Build a mapping from subgrid number to actual {@link SubGridContainer}.
   *
   * 2- and 3-winding transformers are handled in the sense that they are always placed
   * in the lower subgrid of the two grids they are connecting.
   *
   * As transformers have their nodes in two different subgrids, all real nodes of the transformers
   * are duplicated and placed in the upper and in the lower subgrid as part of the grids
   * {@link RawGridElements#nodes} set. The high voltage nodes of these transformer nodes are
   * NOT marked as slack nodes yet, if they haven't been marked as slacks before.
   *
   * To create fully calculable {@link SubGridContainer} with slack nodes in each subgrid 
   * use // todo THIS METHOD TO TRANSFORM ALL HIGHER VOLTAGE NODES OF THE TRANSFORMERS IN THE SUBGRID 
   *   TO BE SLACK NODES 
   *
   * @param gridName Name of the grid
   * @param subnetNumbers Set of available subne numbers
   * @param rawGrid Container model with all raw grid elements
   * @param systemParticipants Container model with all system participant inputs
   * @param graphics Container model with all graphic elements
   * @return A mapping from subnet number to container model with sub grid elements
   */
```

### PR DESCRIPTION
**Pull request by johanneshiry**
Resolves #124 

- I'm actually not quite happy with the name of the method. Feel free to propose another name for the method.
- Please check the javadocs carefully as the description is important for further understanding of what happens inside the method

### CODE REVIEW

**Comment by johanneshiry** [comment]
!test

File: src/main/java/edu/ie3/datamodel/utils/ContainerUtils.java [hunk 1]
```diff
@@ -418,4 +418,236 @@ public static JointGridContainer combineToJointGrid(
     return new JointGridContainer(
         gridName, rawGrid, systemParticipants, graphicElements, subGridTopologyGraph);
   }
+
+  /**
+   * Modifies a given {@link SubGridContainer} to make it computable for power flow calculations
+   * based on its structure. This means, that the grid is modified in a way that slack nodes are
+   * added for transformers based on assumptions about the grid, as well as all other affect
+   * entities of the grid are accordingly.
```
Comment (ckittl) [inline src/main/java/edu/ie3/datamodel/utils/ContainerUtils.java:426]: I would give a little more more insight on why this is even necessary. Some suggestion:

```
Returns a copy of the given {@link SubGridContainer} with a certain set of nodes marked as slack nodes. By default, when the container is derived from {@link JointGridContainer}, only the original slack nodes are incorporated in the different sub containers. Thereby, power flow calculations of most of them cannot be carried out right away. This means, that the grid is modified in a way that slack nodes are added at transformer nodes based on assumptions about the grid, as well as all other affect entities of the grid are accordingly.
```

File: src/main/java/edu/ie3/datamodel/utils/ContainerUtils.java [hunk 2]
```diff
@@ -418,4 +418,236 @@ public static JointGridContainer combineToJointGrid(
...
+   *       <ul>
+   *         <li>high voltage nodes are marked as slack nodes
+   *         <li>high voltage nodes in the {@link RawGridElements#getNodes()} set are replaced with
+   *             the new slack marked high voltage nodes
+   *         <li>high voltage nodes as part of {@link GraphicElements#getNodeGraphics()} are
+   *             replaced with the new slack marked high voltage nodes
+   *       </ul>
+   *   <li>3 winding transformer handling
+   *       <ul>
+   *         <li>if node a is located in this grid, no changes on 3 winding transformer nodes are
```
Comment (ckittl) [inline src/main/java/edu/ie3/datamodel/utils/ContainerUtils.java:441]:
```suggestion
   *         <li>if node a is located in this sub grid, no changes on 3 winding transformer nodes are
```

File: src/main/java/edu/ie3/datamodel/utils/ContainerUtils.java [hunk 3]
```diff
@@ -418,4 +418,236 @@ public static JointGridContainer combineToJointGrid(
...
+   *             made
+   *         <li>if node b or c is located in this grid, the transformers internal node is marked as
+   *             slack node and if node a is marked as slack node, this node is unmarked as slack
+   *             node
```
Comment (ckittl) [inline src/main/java/edu/ie3/datamodel/utils/ContainerUtils.java:445]: Pretty good! Didn't see this point! 👍

File: src/main/java/edu/ie3/datamodel/utils/ContainerUtils.java [hunk 4]
```diff
@@ -418,4 +418,236 @@ public static JointGridContainer combineToJointGrid(
...
+   *         <li>if node a got unmarked as slack, the {@link RawGridElements#getNodes()} gets
+   *             adapted accordingly
+   *         <li>in any case the internal node of the transformer is added to the {@link
+   *             RawGridElements#getNodes()} set
+   *         <li>the resulting 3 winding transformer allows for either a full calculation of this
+   *             transformer model (just ignore the internal node) or a transfer to a
+   *             'close-to-T-equivalent' model that allows for a simplified calculation of the
+   *             transformer power flows
```
Comment (ckittl) [inline src/main/java/edu/ie3/datamodel/utils/ContainerUtils.java:453]: I don't know, if this is not confusing people, as it is obvious, that you can do different stuff with the transformer model.

File: src/main/java/edu/ie3/datamodel/utils/ContainerUtils.java [hunk 5]
```diff
@@ -418,4 +418,236 @@ public static JointGridContainer combineToJointGrid(
...
+   *       </ul>
+   * </ul>
+   *
+   * @param subGridContainer the subgrid container that should be modified
```
Comment (ckittl) [inline src/main/java/edu/ie3/datamodel/utils/ContainerUtils.java:457]:
```suggestion
   * @param subGridContainer the subgrid container to be altered
```

File: src/main/java/edu/ie3/datamodel/utils/ContainerUtils.java [hunk 6]
```diff
@@ -418,4 +418,236 @@ public static JointGridContainer combineToJointGrid(
...
+   * @return a modified, power flow computable subgrid container
```
Comment (ckittl) [inline src/main/java/edu/ie3/datamodel/utils/ContainerUtils.java:458]:
```suggestion
   * @return a copy of the given {@link SubGridContainer} with transformer nodes marked as slack
```

File: src/main/java/edu/ie3/datamodel/utils/ContainerUtils.java [hunk 7]
```diff
@@ -418,4 +418,236 @@ public static JointGridContainer combineToJointGrid(
...
+   */
+  public static SubGridContainer computableSubGrid(final SubGridContainer subGridContainer) {
```
Comment (ckittl) [inline src/main/java/edu/ie3/datamodel/utils/ContainerUtils.java:460]:
```suggestion
  public static SubGridContainer withTransformerNodesAsSlack(final SubGridContainer subGridContainer) {
```
What do you think?

File: src/main/java/edu/ie3/datamodel/utils/ContainerUtils.java [hunk 8]
```diff
@@ -418,4 +418,236 @@ public static JointGridContainer combineToJointGrid(
...
+                Collectors.toMap(
+                    AbstractMap.SimpleEntry::getKey, AbstractMap.SimpleEntry::getValue));
+
+    // build the updated 2w transformer with nodeA as slack and add it to our new set
+    Set<Transformer2WInput> newTrafos2w =
+        subGridContainer.getRawGrid().getTransformer2Ws().stream()
+            .map(
+                oldTrafo2w ->
+                    copyTrafo2WUpdateNodeA(
+                        oldTrafo2w, oldToNewTrafo2WNodes.get(oldTrafo2w.getNodeA())))
```
Comment (ckittl) [inline src/main/java/edu/ie3/datamodel/utils/ContainerUtils.java:527]:
```suggestion
                    copyTrafo2WUpdateNodeA(
                        oldTrafo2w, oldToNewTrafo2WNodes.getOrElse(oldTrafo2w.getNodeA(), throw new ElementNotFoundException("<place some meaningful message here>"))))
```

File: src/main/java/edu/ie3/datamodel/utils/ContainerUtils.java [hunk 9]
```diff
@@ -418,4 +418,236 @@ public static JointGridContainer combineToJointGrid(
...
+            .collect(Collectors.toSet());
+
+    // update node input graphics (2 winding transformers only)
```
Comment (ckittl) [inline src/main/java/edu/ie3/datamodel/utils/ContainerUtils.java:530]: Wouldn't it also affect the nodes "A", of a three winding transformer, that are set to `slack = false`?

File: src/main/java/edu/ie3/datamodel/utils/ContainerUtils.java [hunk 10]
```diff
@@ -418,4 +418,236 @@ public static JointGridContainer combineToJointGrid(
...
+    Set<NodeGraphicInput> newNodeGraphics =
+        Stream.concat(
+                // filter old ones
+                subGridContainer.getGraphics().getNodeGraphics().stream()
+                    .filter(nodeGraphic -> !oldToNewNodeGraphics.containsKey(nodeGraphic)),
+                // add the new trafo2w ones
+                oldToNewNodeGraphics.values().stream())
+            .collect(Collectors.toSet());
+
+    // update nodes in raw grid by removing the old 2 winding transformer nodes and add all new ones
```
Comment (ckittl) [inline src/main/java/edu/ie3/datamodel/utils/ContainerUtils.java:555]:
```suggestion
    // update nodes in raw grid by removing the old transformer nodes and add all new ones
```

File: src/main/java/edu/ie3/datamodel/utils/ContainerUtils.java [hunk 11]
```diff
@@ -418,4 +418,236 @@ public static JointGridContainer combineToJointGrid(
...
+    Set<NodeInput> newNodes =
+        Stream.concat(
+                // filter the old ones (trafo2w, trafo3wToBeRemoved)
+                subGridContainer.getRawGrid().getNodes().stream()
+                    .filter(node -> !oldToNewTrafo2WNodes.containsKey(node))
+                    .filter(node -> !oldToNewTrafo3WNodes.containsKey(node)),
+                // add the new ones (trafo2w, trafo3w and updated trafo3w nodeA (previous slacks))
```
Comment (ckittl) [inline src/main/java/edu/ie3/datamodel/utils/ContainerUtils.java:562]:
```suggestion
                // add the new ones (trafo2w, trafo3w internal and updated trafo3w nodeA (previous slacks))
```

File: src/main/java/edu/ie3/datamodel/utils/ContainerUtils.java [hunk 12]
```diff
@@ -418,4 +418,236 @@ public static JointGridContainer combineToJointGrid(
...
+   * </ul>
+   *
+   * @param subGridContainer the subgrid container that should be modified
+   * @return a modified, power flow computable subgrid container
+   */
+  public static SubGridContainer computableSubGrid(final SubGridContainer subGridContainer) {
+
+    // transformer 3w
+    Map<NodeInput, NodeInput> oldToNewTrafo3WNodes = new HashMap<>();
+    Map<Transformer3WInput, NodeInput> newTrafos3w =
```
Comment (ckittl) [inline src/main/java/edu/ie3/datamodel/utils/ContainerUtils.java:464]:
```suggestion
    Map<Transformer3WInput, NodeInput> transformers3wToInternalNode =
```

Further down the file, it is not very comprehensible, why `newTrafos3w.values()` are added to a set of nodes.

File: src/main/java/edu/ie3/datamodel/utils/ContainerUtils.java [hunk 13]
```diff
@@ -418,4 +418,236 @@ public static JointGridContainer combineToJointGrid(
...
+    return new SubGridContainer(
+        subGridContainer.getGridName(),
+        subGridContainer.getSubnet(),
+        rawGridCopy4Slacks(
+            subGridContainer.getRawGrid(), newNodes, newTrafos2w, newTrafos3w.keySet()),
+        subGridContainer.getSystemParticipants(),
+        graphicsCopy4Slacks(subGridContainer.getGraphics(), newNodeGraphics));
+  }
+
+  private static Transformer3WInput copyTrafo3WWithInternalSlack(
```
Comment (ckittl) [inline src/main/java/edu/ie3/datamodel/utils/ContainerUtils.java:578]:
```suggestion
  private static Transformer3WInput copyTransformer3WInputWithInternalSlack(
```
Maybe to stick a little more to the model class? Would you mind adding JavaDoc?

File: src/main/java/edu/ie3/datamodel/utils/ContainerUtils.java [hunk 14]
```diff
@@ -418,4 +418,236 @@ public static JointGridContainer combineToJointGrid(
...
+      NodeGraphicInput oldNodeGraphic, NodeInput newNodeInput) {
+    return new NodeGraphicInput(
+        oldNodeGraphic.getUuid(),
+        oldNodeGraphic.getGraphicLayer(),
+        oldNodeGraphic.getPath(),
+        newNodeInput,
+        oldNodeGraphic.getPoint());
+  }
+
+  private static RawGridElements rawGridCopy4Slacks(
```
Comment (ckittl) [inline src/main/java/edu/ie3/datamodel/utils/ContainerUtils.java:606]: Please add JavaDoc here.

File: src/main/java/edu/ie3/datamodel/utils/ContainerUtils.java [hunk 15]
```diff
@@ -418,4 +418,236 @@ public static JointGridContainer combineToJointGrid(
...
+    return new RawGridElements(
+        newNodes,
+        rawGridElements.getLines(),
+        newTrafos2w,
+        newTrafos3w,
+        rawGridElements.getSwitches(),
+        rawGridElements.getMeasurementUnits());
+  }
+
+  private static GraphicElements graphicsCopy4Slacks(
```
Comment (ckittl) [inline src/main/java/edu/ie3/datamodel/utils/ContainerUtils.java:621]: Here als well.

File: src/test/groovy/edu/ie3/datamodel/io/source/csv/CsvTestDataMeta.groovy [hunk 16]
```diff
@@ -7,11 +7,7 @@ package edu.ie3.datamodel.io.source.csv
 
 import edu.ie3.datamodel.io.FileNamingStrategy
 
-/**
- * //ToDo: Class Description
- *
- * @version 0.1* @since 13.04.20
- */
+
```
Comment (ckittl) [inline src/test/groovy/edu/ie3/datamodel/io/source/csv/CsvTestDataMeta.groovy:10]: Nice move. 😎 😅

File: src/test/groovy/edu/ie3/datamodel/utils/ContainerUtilsTest.groovy [hunk 17]
```diff
@@ -168,6 +171,71 @@ class ContainerUtilsTest extends Specification {
 		}
 	}
 
+	def "The container util modifies provided subgrids to make them computable as expected"() {
```
Comment (ckittl) [inline src/test/groovy/edu/ie3/datamodel/utils/ContainerUtilsTest.groovy:174]:
```suggestion
	def "The container util returns copy of provided subgrids with slack nodes marked as expected"() {
```

File: src/test/groovy/edu/ie3/datamodel/utils/ContainerUtilsTest.groovy [hunk 18]
```diff
@@ -168,6 +171,71 @@ class ContainerUtilsTest extends Specification {
...
+					}
+				}
+
+				// all adapted trafo2w nodes must be part of the nodes set
+				assert rawGrid.nodes.containsAll(rawGrid.transformer2Ws.collect{it.nodeA})
+
+				// 3 winding transformer slack nodes must be mapped correctly
+				rawGrid.transformer3Ws.each {
+					def trafo3w = it
+					if(trafo3w.nodeA.subnet == subnet){
```
Comment (ckittl) [inline src/test/groovy/edu/ie3/datamodel/utils/ContainerUtilsTest.groovy:218]:
```suggestion
					if(trafo3w.nodeA.subnet == subnet) {
```

File: src/test/groovy/edu/ie3/datamodel/utils/ContainerUtilsTest.groovy [hunk 19]
```diff
@@ -168,6 +171,71 @@ class ContainerUtilsTest extends Specification {
...
+						// subnet 1 is highest grid in test set + trafo 3w -> nodeA must be slack
+						assert subnet == 1 ? trafo3w.nodeA.slack : !trafo3w.nodeA.slack
+						assert !trafo3w.nodeInternal.slack
+						assert rawGrid.nodes.contains(trafo3w.nodeInternal)
+					}else{
```
Comment (ckittl) [inline src/test/groovy/edu/ie3/datamodel/utils/ContainerUtilsTest.groovy:223]:
```suggestion
					} else {
```

**Comment by sensarmad** [comment]
suggestion for the method name: assumeTrafoNodesAsSlack

File: src/main/java/edu/ie3/datamodel/utils/ContainerUtils.java (hunk 4 above)
Comment (johanneshiry) [reply src/main/java/edu/ie3/datamodel/utils/ContainerUtils.java:453]: you're right - removed it

File: src/main/java/edu/ie3/datamodel/utils/ContainerUtils.java (hunk 8 above)
Comment (johanneshiry) [reply src/main/java/edu/ie3/datamodel/utils/ContainerUtils.java:527]: ... when scala code appears in java 😏

File: src/main/java/edu/ie3/datamodel/utils/ContainerUtils.java (hunk 9 above)
Comment (johanneshiry) [reply src/main/java/edu/ie3/datamodel/utils/ContainerUtils.java:530]: yep but I thought as we never drawn 3 winding transformers yet, there is no need to adapt this here - added the corresponding code though

File: src/test/groovy/edu/ie3/datamodel/io/source/csv/CsvTestDataMeta.groovy (hunk 16 above)
Comment (johanneshiry) [reply src/test/groovy/edu/ie3/datamodel/io/source/csv/CsvTestDataMeta.groovy:10]: 😎

**Pull request merged**